File: app/src/main/java/d/spidchenko/flashcards/ui/main/MainViewModel.java
package d.spidchenko.flashcards.ui.main;

import static d.spidchenko.flashcards.Dictionary.translations;

import android.app.Application;
import android.util.Log;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

import d.spidchenko.flashcards.Word;
import d.spidchenko.flashcards.db.DatabaseHelper;

public class MainViewModel extends AndroidViewModel {

    private static final String TAG = "MainViewModel.LOG_TAG";
    public static final String KEY = "words";
    private final Application mApplication;
    private ArrayList<Word> mWords;
    private int mCurrentIdx;
    private Word mCurrentWord;
    private final MutableLiveData<String> mCurrentTranslation = new MutableLiveData<>();
    private MutableLiveData<Boolean> mIsSpeechSynthesizerEnabled = new MutableLiveData<>();


    public MainViewModel(Application application) {
        super(application);
        this.mApplication = application;
        initDatabaseWithWords();
        getAllWords();
        mIsSpeechSynthesizerEnabled.setValue(true);
    }

    LiveData<String> getCurrentTranslation() {
        return mCurrentTranslation;
    }

    LiveData<Boolean> isSpeechSynthesizerEnabled(){
        return mIsSpeechSynthesizerEnabled;
    }

    public void nextWord() {
        if (mCurrentIdx > mWords.size() - 1) {
            mCurrentIdx = 0;
            mWords.sort(Comparator.comparingInt(Word::getMemoryRate));
        }
        mCurrentWord = mWords.get(mCurrentIdx++);
        Log.d(TAG, "getNextWord: idx=" + mCurrentIdx + " " + mCurrentWord);
        mCurrentTranslation.postValue(mCurrentWord.getRuWord());
    }

    public void translate() {
        if (Objects.equals(mCurrentTranslation.getValue(), mCurrentWord.getRuWord())) {
            mCurrentTranslation.setValue(mCurrentWord.getPlWord());
        } else {
            mCurrentTranslation.setValue(mCurrentWord.getRuWord());
        }
    }

    public void toggleSpeechSynthesizerState(){
        if (mIsSpeechSynthesizerEnabled.getValue() != null){
            mIsSpeechSynthesizerEnabled.setValue(!mIsSpeechSynthesizerEnabled.getValue());
        }
    }

    public void increaseRate() {
        if (mCurrentWord != null) {
            int rate = mCurrentWord.getMemoryRate();
            mCurrentWord.setMemoryRate(++rate);
            updateWord(mCurrentWord);
        }
    }

    public void decreaseRate() {
        if (mCurrentWord != null) {
            int rate = mCurrentWord.getMemoryRate();
            mCurrentWord.setMemoryRate(--rate);
            updateWord(mCurrentWord);
        }
    }

    private void updateWord(Word word) {
        new Thread(() -> {
            DatabaseHelper db = DatabaseHelper.getInstance(mApplication);
            db.updateWord(word);
            db.close();
        }).start();
    }

    private void getAllWords() {
        new Thread(() -> {
            DatabaseHelper db = DatabaseHelper.getInstance(mApplication);
            List<Word> words = db.getAllWords();
            db.close();
            mWords = new ArrayList<>(words);
            shuffleWordsButSaveRateOrder();
            Log.d(TAG, "getAllWords: Total words in db " + words.size());
            nextWord();
        }).start();
    }

    private void shuffleWordsButSaveRateOrder(){
        Collections.shuffle(mWords);
        mWords.sort(Comparator.comparingInt(Word::getMemoryRate));
    }

    private void saveNewWord(Word word) {
        new Thread(() -> {
            DatabaseHelper db = DatabaseHelper.getInstance(mApplication);
            db.addWord(word);
            Log.d(TAG, "saveNewWord: " + word.getRuWord());
        }).start();
    }

    private void saveAllWords(List<Word> words) {
        new Thread(() -> {
            DatabaseHelper db = DatabaseHelper.getInstance(mApplication);
            db.addAllWords(words);
//            Log.d(TAG, "saveAllWords: " + words);
        }).start();
    }

    private void initDatabaseWithWords() {
        List<Word> words = new ArrayList<>();
        for (int i = 0; i < translations.length; i += 2) {
//            Log.d(TAG, "initDatabaseWithWords: " +
//                    "ru{" + translations[i] + "} pl{" + translations[i + 1] + "}");
            if ("".equals(translations[i])) {
                break;
            }
            words.add(new Word(translations[i], translations[i + 1]));
        }
        saveAllWords(words);
    }

}

File: app/src/main/java/d/spidchenko/flashcards/ui/main/MainViewModel.kt
package d.spidchenko.flashcards.ui.main

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import d.spidchenko.flashcards.data.Dictionary.translations
import d.spidchenko.flashcards.db.dao.WordDao
import d.spidchenko.flashcards.db.models.Word
import d.spidchenko.flashcards.tts.VoiceSynthesizer


class MainViewModel(
    private val wordDao: WordDao,
//    private val mDatabaseHelper: DatabaseHelper,
    private val mVoiceSynthesizer: VoiceSynthesizer
) : ViewModel() {
    private lateinit var words: ArrayList<Word>
    private var currentIdx = 0
    private lateinit var currentWord: Word
    private val mCurrentTranslation = MutableLiveData<String>()
    private val mIsSpeechSynthesizerEnabled = MutableLiveData<Boolean>()
    val currentTranslation: LiveData<String>
        get() = mCurrentTranslation
    val isSpeechSynthesizerEnabled: LiveData<Boolean>
        get() = mIsSpeechSynthesizerEnabled

    fun nextWord() {
        if (currentIdx > words.size - 1) {
            currentIdx = 0
            words.sortWith(Comparator.comparingInt(Word::memoryRate))
        }
        currentWord = words[currentIdx++]
        Log.d(TAG, "getNextWord: idx=$currentIdx $currentWord")
        mCurrentTranslation.postValue(currentWord.ruWord)
    }

    fun translate() {
        if (mCurrentTranslation.value == currentWord.ruWord) {
            mCurrentTranslation.value = currentWord.plWord
            if (java.lang.Boolean.TRUE == mIsSpeechSynthesizerEnabled.value) {
                mVoiceSynthesizer.speak(currentWord.plWord)
            }
        } else {
            mCurrentTranslation.setValue(currentWord.ruWord)
        }
    }

    fun toggleSpeechSynthesizerState() {
        if (mIsSpeechSynthesizerEnabled.value != null) {
            mIsSpeechSynthesizerEnabled.value = !mIsSpeechSynthesizerEnabled.value!!
        }
    }

    fun increaseRate() {
        var rate = currentWord.memoryRate
        currentWord.memoryRate = ++rate
        updateWord(currentWord)
    }

    fun decreaseRate() {
        var rate = currentWord.memoryRate
        currentWord.memoryRate = --rate
        updateWord(currentWord)
    }

    private fun updateWord(word: Word) {
        Thread {
            wordDao.updateWords(word)
//            mDatabaseHelper.close()
        }.start()
    }

    private fun getAllWords() {
        Thread {
            val words: List<Word> = wordDao.getAllWords()
//            mDatabaseHelper.close()
            this.words = ArrayList(words)
            shuffleWordsButSaveRateOrder()
            Log.d(TAG, "getAllWords: Total words in db " + words.size)
            nextWord()
        }.start()
    }

    private fun shuffleWordsButSaveRateOrder() {
        words.shuffle()
        words.sortWith(Comparator.comparingInt(Word::memoryRate))
    }

    private fun saveNewWord(word: Word) {
        Thread {
//            wordDao.insertAll(word)
            Log.d(TAG, "saveNewWord: " + word.ruWord)
        }.start()
    }

    private fun saveAllWords(words: List<Word>) {
        Thread { wordDao.insertAll(words)}.start()
    }

    private fun initDatabaseWithWords() {
        val words = mutableListOf<Word>()
        var i = 0
        while (i < translations.size) {
            if (translations[i].isBlank()) {
                break
            }
            words.add(Word(translations[i], translations[i + 1]))
            i += 2
        }
        saveAllWords(words)
    }

    companion object {
        private const val TAG = "MainViewModel.LOG_TAG"
        const val KEY = "words"
    }

    init {
//        initDatabaseWithWords()
        getAllWords()
        mIsSpeechSynthesizerEnabled.value = true
    }
}